examples/lfric/scripts/gpu_offloading.py

```diff
@@ -0,0 +1,187 @@
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+# COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Authors: A. R. Porter, STFC Daresbury Lab
+#          R. W. Ford, STFC Daresbury Lab
+#          S. Siso, STFC Daresbury Lab
+#          L. Mosimann, NVIDIA.
+
+'''PSyclone transformation script for LFRic to apply colouring and GPU
+offloading. Also adds redundant computation to the level-1 halo for
+setval_* generically.
+
+'''
+import os
+from psyclone.domain.lfric import LFRicConstants
+from psyclone.psyir.nodes import Directive, Loop
+from psyclone.psyir.transformations import (
+    ACCKernelsTrans, TransformationError, OMPTargetTrans)
+from psyclone.transformations import (
+    Dynamo0p3ColourTrans, Dynamo0p3OMPLoopTrans,
+    Dynamo0p3RedundantComputationTrans, OMPParallelTrans,
+    ACCParallelTrans, ACCLoopTrans, ACCRoutineTrans,
+    OMPDeclareTargetTrans, OMPLoopTrans)
+
+
+# Names of any invoke that we won't add any GPU offloading
+INVOKE_EXCLUSIONS = [
+]
+
+OFFLOAD_USING_OMP = os.getenv('OFFLOAD_USING_OMP', False)
+
+
+def trans(psy):
+    '''Applies PSyclone colouring and GPU offloading transformations. Any
+    kernels that cannot be offloaded to GPU are parallelised using OpenMP
+    on the CPU. Any setval_* kernels are transformed so as to compute
+    into the L1 halos.
+
+    '''
+    rtrans = Dynamo0p3RedundantComputationTrans()
+    ctrans = Dynamo0p3ColourTrans()
+    otrans = Dynamo0p3OMPLoopTrans()
+    const = LFRicConstants()
+    cpu_parallel = OMPParallelTrans()
+
+    if OFFLOAD_USING_OMP:
+        # Use OpenMP offloading
+        loop_offloading_trans = OMPLoopTrans(
+            omp_directive="teamsdistributeparalleldo",
+            omp_schedule="none"
+        )
+        kernels_trans = None
+        gpu_region_trans = OMPTargetTrans()
+        gpu_annotation_trans = OMPDeclareTargetTrans()
+    else:
+        # Use OpenACC offloading
+        loop_offloading_trans = ACCLoopTrans()
+        kernels_trans = ACCKernelsTrans()
+        gpu_region_trans = ACCParallelTrans(default_present=False)
+        gpu_annotation_trans = ACCRoutineTrans()
```

Comment:
@arporter this is ready for review, unfortunately github is not picking up that I renamed acc_parellel.py to gpu_offloading.py probably because I also made some changes to make the language more neutral instead of ACC specific and added some prints and comments, but the main change comes from the selected code.

I used the OFFLOAD_USING_OMP env variable here and in the nvfortran.mk initially to not modify the default OpenACC behaviour but I notice that with the updated nvfortran.mk there is no way to compile with nvfortran wihtout offloading, another alternative would be to have an OFFLOAD_LANGUAGE=[omp|acc] env var which this script will complain if its not one of these values but nvfortran.mk do the right thing.

There are other alternatives to envvars like replicating the script with the other language or using a common utils, but since we also need a mechanism to select the right flags anyway I though the envvar approach was the easiest. Let me know what you think.